include/fi_util.h

```diff
@@ -292,6 +293,62 @@ int ofi_get_src_addr(uint32_t addr_format,
 		     const void *dest_addr, size_t dest_addrlen,
 		     void **src_addr, size_t *src_addrlen);
 void ofi_getnodename(char *buf, int buflen);
+int ofi_av_get_index(struct util_av *av, const void *addr);
+
+/*
+ * Connection Map
+ */
+
+enum util_cmap_state {
+	CMAP_UNSPEC,
+	CMAP_CONNECTING,
+	CMAP_CONNECTED
+};
+
+struct util_cmap_handle {
+	struct util_cmap *cmap;
+	enum util_cmap_state state;
+	struct util_cmap_key *key;
+	size_t key_index;
+	fi_addr_t fi_addr;
+	struct util_cmap_peer *peer;
+};
+
+struct util_cmap_key {
+	struct util_cmap_handle *handle;
+};
+DECLARE_FREESTACK(struct util_cmap_key, util_cmap_keypool);
+
+struct util_cmap_peer {
+	struct util_cmap_handle *handle;
+	struct dlist_entry entry;
+	size_t addrlen;
+	uint8_t addr[];
+};
+
+typedef void (*ofi_cmap_free_handle_func)(void *arg);
+
+struct util_cmap {
+	struct util_av *av;
+	struct util_cmap_handle **handles;
+	struct util_cmap_keypool *keypool;
+	struct dlist_entry peer_list;
+	ofi_cmap_free_handle_func free_handle;
+	fastlock_t lock;
+};
+
+void ofi_cmap_update_state(struct util_cmap_handle *handle,
+		enum util_cmap_state state);
+int ofi_cmap_add_handle_addr(struct util_cmap *cmap, struct util_cmap_handle *handle,
```

Comment:
The two similarly named functions take in different arguments to add a handle. They could be merged with the presence/absence of the args defining the behavior. That would reduce the number of functions exported. Also, ofi_cmap_del_handles is not supposed to be exported. I shall remove it from the header.

> the "get" function does deletion

This function moves the handle from "peer list" to handle array in cmap which appears like deletion. An alternate approach would be to put the move part in a separate function which should be called when a provider does an av_insert.
> For those functions that require a lock being held, please also add that information to the header file.

Sure, will do. Thank you for reviewing.